### ISSUE TITLE
MSSQL: Unique index with filter is not created (Sequelize v4)

### ISSUE THREAD

**Issue by david-bojnansky** [sequelize/sequelize#7746]
`Sequelize [Node: 6.10.1, CLI: 2.7.0, ORM: 4.0.0]
`
I tried to create two tables and both with one unique index with filter where X IS NOT NULL. One table is created as expected. But the other one is not. I am using TypeScript, but I will post plain JS codes of both models.

**Article Model**

```js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Sequelize = require("sequelize");
const db_1 = require("../../db");
exports.Article = db_1.iisDb.define('article', {
    s4_id: {
        type: Sequelize.UUID,
        allowNull: true,
        //  unique: true,
        comment: 'Artikly_Artikl.ID',
    },
    name: {
        type: Sequelize.STRING(100),
        allowNull: true,
    },
    s4_name: {
        type: Sequelize.STRING(100),
        allowNull: true,
        comment: 'Artikly_Artikl.Nazev',
    },
}, {
    paranoid: true,
    indexes: [
        {
            name: 'articles_s4_id_unique', // also tried without name and worked too, in the second case not...
            unique: true,
            fields: ['s4_id'],
            where: { s4_id: { $ne: null } },
        },
    ],
    instanceMethods: {},
});
exports.default = exports.Article;

```
This one works fine.

```
**Executing (default):**  IF OBJECT_ID('[articles]', 'U') IS NULL CREATE TABLE [articles] ([id] INTEGER NOT NULL IDENTITY(1,1) , [s4_id] CHAR(36) NULL, [name] NVARCHAR(100) NULL, [s4_n
ame] NVARCHAR(100) NULL, [createdAt] DATETIMEOFFSET NOT NULL, [updatedAt] DATETIMEOFFSET NOT NULL, [deletedAt] DATETIMEOFFSET NULL, PRIMARY KEY ([id]));
**Executing (default):**  EXEC sys.sp_helpindex @objname = N'[articles]';
**Executing (default):**  CREATE UNIQUE INDEX [articles_s4_id_unique] ON [articles] ([s4_id]) WHERE [s4_id] IS NOT NULL
```

**Refueling Model**

```js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Sequelize = require("sequelize");
const db_1 = require("../../db");
const refueling_evidence_1 = require("./refueling-evidence");
const tank_1 = require("./tank");
const property_1 = require("./property");
exports.Refueling = db_1.iisDb.define('refueling', {
    s4_id: {
        type: Sequelize.UUID,
        allowNull: true,
        //  unique: true,
        comment: 'Fakturace_PolozkaFakturyPrijate.ID',
    },
    date: {
        type: Sequelize.DATE,
        allowNull: true,
    },
    gain: {
        type: Sequelize.INTEGER,
        allowNull: true,
    },
    s4_gain: {
        type: Sequelize.DECIMAL(18, 4),
        allowNull: true,
        comment: 'Fakturace_PolozkaFakturyPrijate.Mnozstvi',
    },
    note: {
        type: Sequelize.TEXT,
        allowNull: true,
    },
    s4_note: {
        type: Sequelize.TEXT,
        allowNull: true,
        comment: 'Fakturace_PolozkaFakturyPrijate.Poznamka',
    },
}, {
    paranoid: true,
    indexes: [
        {
            name: 'refuelings_s4_id_unique', // also tried without this name; the result was that it created two unique indexes (one without filter and one with filter)
            unique: true,
            fields: ['s4_id'],
            where: { s4_id: { $ne: null } },
        },
    ],
    instanceMethods: {},
});
exports.Refueling.RefuelingEvidence = exports.Refueling.belongsTo(refueling_evidence_1.RefuelingEvidence);
exports.Refueling.Property = exports.Refueling.belongsTo(property_1.Property, {
    foreignKey: { allowNull: false },
});
exports.Refueling.Tank = exports.Refueling.belongsTo(tank_1.Tank);
exports.default = exports.Refueling;
```

This one does not work as the previous one.

```
**Executing (default):** IF OBJECT_ID('[refuelings]', 'U') IS NULL CREATE TABLE [refuelings] ([id] INTEGER NOT NULL IDENTITY(1,1) , [s4_id] CHAR(36) NULL, [date] DATETIMEOFFSET NULL,
[gain] INTEGER NULL, [s4_gain] DECIMAL(18,4) NULL, [note] NVARCHAR(MAX) NULL, [s4_note] NVARCHAR(MAX) NULL, [createdAt] DATETIMEOFFSET NOT NULL, [updatedAt] DATETIMEOFFSET NOT NUL
L, [deletedAt] DATETIMEOFFSET NULL, [refuelingEvidenceId] INTEGER NULL, [propertyId] INTEGER NOT NULL, [tankId] INTEGER NULL, CONSTRAINT [refuelings_s4_id_unique] UNIQUE ([s4_id])
, PRIMARY KEY ([id]), FOREIGN KEY ([refuelingEvidenceId]) REFERENCES [refuelingEvidences] ([id]) ON DELETE SET NULL, FOREIGN KEY ([propertyId]) REFERENCES [properties] ([id]) ON D
ELETE NO ACTION, FOREIGN KEY ([tankId]) REFERENCES [tanks] ([id]) ON DELETE SET NULL);
**Executing (default):** EXEC sys.sp_helpindex @objname = N'[refuelings]';
```

Output when the index NAME was removed:
```
**Executing (default):**  IF OBJECT_ID('[refuelings]', 'U') IS NULL CREATE TABLE [refuelings] ([id] INTEGER NOT NULL IDENTITY(1,1) , [s4_id] CHAR(36) NULL UNIQUE, [date] DATETIMEOFFSET
 NULL, [gain] INTEGER NULL, [s4_gain] DECIMAL(18,4) NULL, [note] NVARCHAR(MAX) NULL, [s4_note] NVARCHAR(MAX) NULL, [createdAt] DATETIMEOFFSET NOT NULL, [updatedAt] DATETIMEOFFSET
NOT NULL, [deletedAt] DATETIMEOFFSET NULL, [refuelingEvidenceId] INTEGER NULL, [propertyId] INTEGER NOT NULL, [tankId] INTEGER NULL, CONSTRAINT [refuelings_s4_id_unique] UNIQUE ([
s4_id]), PRIMARY KEY ([id]), FOREIGN KEY ([refuelingEvidenceId]) REFERENCES [refuelingEvidences] ([id]) ON DELETE SET NULL, FOREIGN KEY ([propertyId]) REFERENCES [properties] ([id
]) ON DELETE NO ACTION, FOREIGN KEY ([tankId]) REFERENCES [tanks] ([id]) ON DELETE SET NULL);
**Executing (default):**  EXEC sys.sp_helpindex @objname = N'[refuelings]';
**Executing (default):**  CREATE UNIQUE INDEX [refuelings_s4_id] ON [refuelings] ([s4_id]) WHERE [s4_id] IS NOT NULL
```
In this case two unique indexes were created. One without filter and one with filter.

The models are synced in migration:
```js
exports.up = (qi, Sequelize) => {
  return Article.sync()
      .then(() => Property.sync())
      .then(() => Tank.sync())
      .then(() => Machine.sync())
      .then(() => RefuelingEvidence.sync())
      .then(() => Refueling.sync())
      .then(() => Person.sync())
      .then(() => User.sync());
};
```

**Comment by harshithkashyap** [comment]
@david-bojnansky You've mentioned that its working in one model but not the other. Is it possible that previous index for the second model is never cleared from the db? This seems like an issue from your end.
Also, you are mixing up models and migrations. You should be using `queryInterface.createTable` instead of `sync` to create initial set of tables.

**Comment by david-bojnansky** [comment]
harshithkashyap, thanks for your response. The previous index is cleared from db for sure, because I can simply roll back the migration and all the tables will be deleted. Then I will run the migration with the syncing models and it will add all the models without problem. Just in Refueling Model the filter is not created (when the name is present) or is created but twice when the name is not present. 
Secondly, yes, I mixing up models and migrations. But I don't see it as a problem. I know there is QueryInterface.createTable() but I do not want to repeat myself. The sync option is great. Later when I will want to alter the table, then I will just QueryInterface. But for init the db tables I think it is ok to use in this form.

**Comment by david-bojnansky** [comment]
Seems like there is problem when model has associations... When I comment all the associations in Refueling Model, then it created index as expected. When I uncomment only one association (e.g. Refuelin.Tank), then the problem occurred again...

**Comment by harshithkashyap** [comment]
@david-bojnansky I'll close the issue then.

Its best to keep the models and migration files separate. Lets say after sometime you add a column using migration, update the same in model, it'll work on machines where previous migrations are already run. In case you start migration on a new db, there'll be a conflict as Model would have the new column and it gets created during first migration which uses `sync`. When it tries to run the second migration, this will fail as the new column is already created.

**Comment by david-bojnansky** [comment]
In this case I can add a condition to add a new column if it not already exists. But that's not important now. Why is it closed? When I run the syncs outside of migration, still there is a problem. This is not about migration...

**Comment by harshithkashyap** [comment]
@david-bojnansky If you observe the create table log, a unique constraint is created.
> [tankId] INTEGER NULL, CONSTRAINT [refuelings_s4_id_unique] UNIQUE ([s4_id])

By default SQL Server creates a unique index when you create a unique constraint. Sequelize checks if the index already exists with the above name and skips it if it exists.

I dug through the codebase and I found that `Model.refreshAttributes()` adds unique indexes to `uniqueKeys` which is used [here](https://github.com/sequelize/sequelize/blob/master/lib/dialects/mssql/query-generator.js#L90-L97) to create the constraint during `sync`. 

This is not causing a problem in the first case as [`refreshAttributes()`](https://github.com/sequelize/sequelize/blob/master/lib/model.js#L829) method is called before we add unique indexes to `rawAttributes` [here](https://github.com/sequelize/sequelize/blob/master/lib/model.js#L845-L855). When you call the association such as `belongsTo/hasMany`, the `refreshAttributes()` method is called again and sequelize creates the `UNIQUE` contraint and internally SQL server creates unique index with the same name. I've reopened the issue and tagged this as bug. You are welcome to submit a PR :)
